flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/HandlerResponse.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.handler;
+
+import org.apache.flink.runtime.rest.messages.ResponseBody;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Simple container for the response of a handler, that contains either a response of type {@code P} if the incoming
+ * request was handled successfully, otherwise it contains an error message and an associated error code.
+ *
+ * @param <P> type of successful response
+ */
+public final class HandlerResponse<P extends ResponseBody> {
```

Comment:
I think it would also be nice to allow the user to be able to specify a callback in the response which will be called when the response has been fully sent in the network. The reason for this is to be able to update metrics like "how long did the transfer take?" or "what is the progress so far (for big messages)". This would require adding the callback in the `HandlerResponse` which implements the `ChannelFutureListener` and in the `AbstractRestHandler.sendResponse()` register that listener with the `lastContentFuture`, if I understand correctly.